### PR TITLE
Adding new tests for Blueprints, that ensures compose blueprints does not trigger double validation.

### LINKED ISSUE #145: Test that controller composition does not trigger double validation
See https://github.com/wemake-services/django-modern-rest/issues/85 for the original issue.

Now, when using `compose_controllers()` we have a new controller. We need to double check that this still works as expected and no double validation happens.

### PR DESCRIPTION
**Pull request by golovindev**
# I have made things!

Added new tests for Blueprints, that ensures compose blueprints does not trigger double validation.

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [ ] I have updated the documentation for the changes I have made

## Related issues

Closes #145

### CODE REVIEW

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 1]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
     assert composed.responses == []
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    assert isinstance(composed.responses, list)
```

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 2]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
     for endpoint, description in composed.api_endpoints.items():
         assert len(description.metadata.responses) == 2, endpoint
+
+
+# Tests specifically for validation edge cases and preventing duplication.
+def test_compose_validation_order_independence() -> None:
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: Please, use `@pytest.parametrize` for this test

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 3]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+    """Ensure validation works regardless of blueprint order."""
+    with pytest.raises(EndpointMetadataError, match='different serializer'):
+        compose_blueprints(_GetBlueprint, _DifferentSerializerBlueprint)
+
+    with pytest.raises(EndpointMetadataError, match='different serializer'):
+        compose_blueprints(_DifferentSerializerBlueprint, _GetBlueprint)
+
+    with pytest.raises(EndpointMetadataError, match='different serializer'):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion

    with pytest.raises(EndpointMetadataError, match='blueprints with different serializer'):
```

Comment (kondratevdev) [reply tests/test_unit/test_routing/test_compose_blueprints.py]: Also, same match message in another places

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 4]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+    # as it tests GET instead of POST.
+    with pytest.raises(EndpointMetadataError) as exc_info:
+        compose_blueprints(_GetBlueprint, _DuplicateGetBlueprint)
+    assert 'GET' in str(exc_info.value)
+
+    # This is different from test_compose_controller_no_endpoints
+    # as it uses _ZeroMethodsBlueprint with _GetBlueprint.
+    with pytest.raises(EndpointMetadataError) as exc_info:
+        compose_blueprints(_ZeroMethodsBlueprint, _GetBlueprint)
+    assert 'at least one' in str(exc_info.value)
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
@pytest.mark.parametrize(
    ('blueprints', 'expected_message_part'),
    [
        pytest.param(
            (_GetBlueprint, _DifferentSerializerBlueprint),
            'different serializer'
        ),
        pytest.param(
            (_GetBlueprint, _DuplicateGetBlueprint),
            'GET'
        ),
        pytest.param(
            (_ZeroMethodsBlueprint, _GetBlueprint),
            'at least one'
        ),
    ],
)
def test_compose_validation_error_messages(blueprints, expected_message_part):
    """Ensure that validation error messages are descriptive and helpful."""
    with pytest.raises(EndpointMetadataError) as exc_info:
        compose_blueprints(*blueprints)
    
    assert expected_message_part in str(exc_info.value)
```

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 5]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+
+
+def test_compose_sync_async_no_method_conflict() -> None:
+    """Ensure sync/async mix raises error even without method conflict."""
+    with pytest.raises(EndpointMetadataError, match='all sync or all async'):
+        compose_blueprints(_GetBlueprint, _PostBlueprint, _AsyncPutBlueprint)
+
+
+def test_compose_options_conflict_with_meta_mixin() -> None:
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: same here

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 6]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+    )
+
+    assert 'OPTIONS' in composed_sync.api_endpoints
+    assert 'GET' in composed_sync.api_endpoints
+
+
+def test_compose_no_double_validation() -> None:
+    """Ensure blueprints are not validated twice in compose_blueprints."""
+
+    class _TrackingValidator1(BlueprintValidator):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    class _FirstValidator(BlueprintValidator):
```

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 7]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+
+        @override
+        def __call__(
+            self,
+            blueprint: type[Blueprint[BaseSerializer]],
+        ) -> None:
+            self.__class__.call_count += 1
+            return super().__call__(blueprint)
+
+    class _TrackingValidator2(BlueprintValidator):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    class _SecondValidator(BlueprintValidator):
```

File: tests/test_unit/test_routing/test_compose_blueprints.py (hunk 6 above)
Comment (kondratevdev) [reply tests/test_unit/test_routing/test_compose_blueprints.py]: Try do not use numbers in class name

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 8]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+
+        @override
+        def __call__(
+            self,
+            blueprint: type[Blueprint[BaseSerializer]],
+        ) -> None:
+            self.__class__.call_count += 1
+            return super().__call__(blueprint)
+
+    class _TestBlueprint1(Blueprint[PydanticSerializer]):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    class _FirstBlueprint(Blueprint[PydanticSerializer]):
```

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 9]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+        blueprint_validator_cls: ClassVar[type[BlueprintValidator]] = (
+            _TrackingValidator1
+        )
+
+        def get(self) -> str:
+            raise NotImplementedError
+
+    class _TestBlueprint2(Blueprint[PydanticSerializer]):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    class _SecondBlueprint(Blueprint[PydanticSerializer]):
```

File: tests/test_unit/test_routing/test_compose_blueprints.py (hunk 6 above)
Comment (kondratevdev) [reply tests/test_unit/test_routing/test_compose_blueprints.py]: Also, you can move this repeatable code into fixture with factory https://docs.pytest.org/en/stable/how-to/fixtures.html#factories-as-fixtures

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 10]
```diff
@@ -149,3 +187,123 @@ def patch(self) -> list[int]:
...
+        blueprint_validator_cls: ClassVar[type[BlueprintValidator]] = (
+            _TrackingValidator2
+        )
+
+        def post(self) -> str:
+            raise NotImplementedError
+
+    initial_count1 = _TrackingValidator1.call_count
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: same here (first, second)

**Comment by kondratevdev** [comment]
Thank you! Please, make some changes :)

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 11]
```diff
@@ -75,9 +117,16 @@ def test_compose_overlapping_controllers() -> None:
...
     """Ensure that controllers with different serializers can't be used."""
-    with pytest.raises(EndpointMetadataError, match='different serializer'):
+    with pytest.raises(
+        EndpointMetadataError,
+        match='blueprints with different serializer',
+    ):
         compose_blueprints(_DifferentSerializerBlueprint, _SyncBlueprint)
-    with pytest.raises(EndpointMetadataError, match='different serializer'):
+
+    with pytest.raises(
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py:144]: You can also parametrize it

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 12]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+        EndpointMetadataError,
+        match='blueprints with different serializer',
+    ):
+        compose_blueprints(*blueprints)
+
+
+@pytest.mark.parametrize(
+    ('blueprints', 'expected_message_part'),
+    [
+        pytest.param(
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: Can we remove `pytest.param`?

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 13]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+
+def test_compose_no_double_validation(
+    make_validated_blueprint: Callable[[str], tuple[type, type]],
+) -> None:
+    """Ensure blueprints are not validated twice in compose_blueprints."""
+    pairs = [make_validated_blueprint('GET'), make_validated_blueprint('POST')]
+
+    initial_counts = tuple(validator.call_count for _, validator in pairs)
+
+    composed = compose_blueprints(*(blueprint for blueprint, _ in pairs))
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]:
```suggestion
    pairs = [make_validated_blueprint('GET'), make_validated_blueprint('POST')]
    initial_counts = tuple(validator.call_count for _, validator in pairs)
    composed = compose_blueprints(*(blueprint for blueprint, _ in pairs))
```

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 14]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+
+    assert 'OPTIONS' in composed.api_endpoints
+    assert 'GET' in composed.api_endpoints
+
+
+def _make_blueprint_class(method: str, validator: type[BlueprintValidator]):
+    """Helper to create a Blueprint class with a single method."""
+    if method == 'GET':
+
+        class _Blueprint(Blueprint[PydanticSerializer]):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: You have blueprint classes at module level, please try use it

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 15]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+            blueprint_validator_cls = validator
+
+            def get(self) -> str:
+                raise NotImplementedError
+
+    elif method == 'POST':
+
+        class _Blueprint(Blueprint[PydanticSerializer]):
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: same here

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 16]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+
+            def __call__(
+                self,
+                blueprint: type[Blueprint[BaseSerializer]],
+            ) -> None:
+                self.__class__.call_count += 1
+                return super().__call__(blueprint)
+
+        blueprint_cls = _make_blueprint_class(method, _Validator)
+        return blueprint_cls, _Validator
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: Let's try to simplify the function. You can just return `_Validator` on demand. And in test function create instance

File: tests/test_unit/test_routing/test_compose_blueprints.py [hunk 17]
```diff
@@ -146,6 +191,143 @@ def patch(self) -> list[int]:
...
+@pytest.mark.parametrize(
+    'blueprints',
+    [
+        (_GetBlueprint, _DifferentSerializerBlueprint),
+        (_DifferentSerializerBlueprint, _GetBlueprint),
+        (_GetBlueprint, _PostBlueprint, _DifferentSerializerBlueprint),
+        (_DifferentSerializerBlueprint, _GetBlueprint, _PostBlueprint),
+    ],
+)
+def test_compose_validation_order_independence(
```
Comment (kondratevdev) [inline tests/test_unit/test_routing/test_compose_blueprints.py]: Why we separate `test_compose_validation_order_independence` and `test_compose_different_serializers`?

**Review by kondratevdev — CHANGES_REQUESTED** [review]

**Comment by kondratevdev** [comment]
Thank you!
